pkg/controller/volume/selinuxwarning/cache/volumecache.go

```diff
@@ -0,0 +1,260 @@
+		if otherPodInfo.changePolicy != changePolicy {
+			// Send conflict to both pods
+			conflicts = append(conflicts, Conflict{
+				PropertyName:       "SELinuxChangePolicy",
+				EventReason:        "SELinuxChangePolicyConflict",
+				Pod:                podKey,
+				PropertyValue:      string(changePolicy),
+				OtherPod:           otherPodKey,
+				OtherPropertyValue: string(otherPodInfo.changePolicy),
+			}, Conflict{
+				PropertyName:       "SELinuxChangePolicy",
+				EventReason:        "SELinuxChangePolicyConflict",
+				Pod:                otherPodKey,
+				PropertyValue:      string(otherPodInfo.changePolicy),
+				OtherPod:           podKey,
+				OtherPropertyValue: string(changePolicy),
+			})
+		}
+		if otherPodInfo.label != label {
+			// Send conflict to both pods
+			conflicts = append(conflicts, Conflict{
+				PropertyName:       "SELinux label",
+				EventReason:        "SELinuxLabelConflict",
+				Pod:                podKey,
+				PropertyValue:      label,
+				OtherPod:           otherPodKey,
+				OtherPropertyValue: otherPodInfo.label,
+			}, Conflict{
+				PropertyName:       "SELinux label",
+				EventReason:        "SELinuxLabelConflict",
+				Pod:                otherPodKey,
+				PropertyValue:      otherPodInfo.label,
+				OtherPod:           podKey,
+				OtherPropertyValue: label,
+			})
+		}
+	}
+	return conflicts
+}
+
+// Remove a pod from the cache. Prunes all empty structures.
+func (c *volumeCache) DeletePod(logger klog.Logger, podKey cache.ObjectName) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	defer c.dump(logger)
+
+	for volumeName, volume := range c.volumes {
+		delete(volume.pods, podKey)
+		if len(volume.pods) == 0 {
+			delete(c.volumes, volumeName)
+		}
+	}
+}
+
+func (c *volumeCache) dump(logger klog.Logger) {
+	if !logger.V(dumpLogLevel).Enabled() {
+		return
+	}
+	logger.Info("VolumeCache dump:")
+
+	// sort the volume to have consistent output
+	volumeIDs := make([]v1.UniqueVolumeName, 0, len(c.volumes))
+	for volumeID := range c.volumes {
+		volumeIDs = append(volumeIDs, volumeID)
+	}
+	sort.Slice(volumeIDs, func(i, j int) bool {
+		return volumeIDs[i] < volumeIDs[j]
+	})
+	for _, volumeID := range volumeIDs {
+		volume := c.volumes[volumeID]
+		logger.Info("Cached volume", "volume", volumeID, "csiDriver", volume.csiDriver)
+
+		// Sort the pods to have consistent output
+		podKeys := make([]cache.ObjectName, 0, len(volume.pods))
+		for podKey := range volume.pods {
+			podKeys = append(podKeys, podKey)
+		}
+		sort.Slice(podKeys, func(i, j int) bool {
+			return podKeys[i].String() < podKeys[j].String()
+		})
+		for _, podKey := range podKeys {
+			podInfo := volume.pods[podKey]
+			logger.Info("  pod", "pod", podKey, "label", podInfo.label, "changePolicy", podInfo.changePolicy)
+		}
+	}
+}
+
+// GetPodsForCSIDriver returns all pods that use volumes with the given CSI driver.
+func (c *volumeCache) GetPodsForCSIDriver(driverName string) []cache.ObjectName {
```

Comment:
Isn't this supposed to work for at least couple of non-CSI drivers too?

Reply:
Evaluation of in-tree volumes are enough at Pod creation time. Only a CSI driver can change its SELinux support in runtime and then we need to re-evaluate all volumes of that driver then.

Reply:
I added a comment there.